# Case Study: Maximizing Profit of a Credit Card Company

# Scenario 1

In [116]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

total_cust=1000 
cust_no_delay=80/100 #Assuming 40% customers paid on time
cust_30_days=14/100 # Assuming 35% paid 30 days after due date
cust_45_days=5/100 # Assuming 16% paid 45 days after due date
cust_60_days=1/100 # Assuming 9% paid 60 days after due date


month_bal=1000

#Revenue per year
mem_fee=12*20*total_cust 
affil_fee=12*10*total_cust 
int_rate=15/100

# Expenses per year
cost_of_funds=12*total_cust*1000*0.065/12
operating_cost=12*25*1000 
loss_rate=3/100

day=30
categories=4
category_type=[]

for cust_category in range(categories):
    if cust_category==0:
        category_type.append('Paid before due date')
    else:
        category_type.append('Paid %d days after due date'%day)
        day+=15     

def interest_calc(payment_days):
    

    average_day_trans= 15
    
    if payment_days==category_type[0]:
        Int_per_card=0

    elif payment_days==category_type[1]:
        Int_per_card=(average_day_trans+15+30)*month_bal*int_rate/365

    elif payment_days==category_type[2]:
        Int_per_card=(average_day_trans+15+45)*month_bal*int_rate/365

    else:
        Int_per_card=(average_day_trans+15+60)*month_bal*int_rate/365
        
    return Int_per_card 
    
def total_int(column):
    
    payment_days=column[0]
    interest_per_card=column[1]
    
    if payment_days==category_type[0]:
        total_interest=0
    elif payment_days==category_type[1]:
        total_interest=12*interest_per_card*cust_30_days*total_cust
    elif payment_days==category_type[2]:
        total_interest=12*interest_per_card*cust_45_days*total_cust
    else:
        total_interest=12*interest_per_card*cust_60_days*total_cust
        
    return total_interest

def PnL_check():
    if net_profit_loss>=0:
        return 'Profit margin in percentage'
    else:
        return 'Loss margin in percentage'
    
Data_Calculations['Estimated_Interest_on_Default']=Data_Calculations['Category'].apply(interest_calc).round(2)
Data_Calculations['Total_Interest_1Year']=Data_Calculations[['Category','Estimated_Interest_on_Default']].apply(total_int,axis=1)
Data_Calculations['Fixed_Late_Fee_charge']=Data_Calculations['Estimated_Interest_on_Default'].apply(lambda x: 0 if x==0 else 50)
Data_Calculations

Category  Estimated_Interest_on_Default  \
1         Paid before due date                           0.00   
2  Paid 30 days after due date                          24.66   
3  Paid 45 days after due date                          30.82   
4  Paid 60 days after due date                          36.99   

   Total_Interest_1Year  Fixed_Late_Fee_charge  
1                   0.0                      0  
2               41428.8                     50  
3               18492.0                     50  
4                4438.8                     50

# Profit Scenario 1

In [117]:
Expenses=cost_of_funds+operating_cost+(loss_rate*total_cust*month_bal)
Earnings=mem_fee+affil_fee+np.sum(Data_Calculations['Total_Interest_1Year'])+(12*max(Data_Calculations['Fixed_Late_Fee_charge'])*5/100*total_cust)

margin=Earnings-Expenses
net_profit_loss=margin/Expenses*100
    
report=pd.DataFrame({'Yearly Card Balance':12*month_bal*total_cust,'Revenue':Earnings,'Expenses':Expenses,
              'Margin':margin,'Profit %':net_profit_loss.round(2)},index=[' '])
report

Yearly Card Balance   Revenue  Expenses   Margin  Profit %
              12000000  454359.6  395000.0  59359.6     15.03

# Scenario 2

In [120]:
def membership_calc(mem_fee,monthly_growth):
    
    monthly_mem_fee=[]
    
    for months in total_months:
        monthly_mem_fee.append(mem_fee)
        new_mem_fee=mem_fee+mem_fee*(monthly_growth)
        mem_fee=new_mem_fee
        
    return monthly_mem_fee


def month_bal_calc(average_bal,monthly_growth):
    
    avg_month_bal=[]
    
    for months in total_months:
        avg_month_bal.append(average_bal)
        new_month_bal=average_bal+average_bal*(monthly_growth)
        average_bal=new_month_bal
        
    return avg_month_bal


total_months=np.arange(1,13) 
mem_fee=20
affil_fee=10 
monthly_growth=0.05
average_bal=1000 
monthly_growth=0.05 
late_fee=50


per_month_fee=membership_calc(mem_fee,monthly_growth)
bal_per_month=month_bal_calc(average_bal,monthly_growth)


Data_Calc_scenario2=pd.DataFrame({'Month':total_months,'Month_on_Month_Bal':bal_per_month,'Membership_Fee':per_month_fee},index=[' ' for index in total_months]).round(2)

#Interest after 'M' months = M*[outstanding balance*(1+Interest rate per year/4)^4-outstanding balance]/12

def interest_calc(month_bal):
    Int_per_card=(month_bal*(1+int_rate/4)**4-month_bal)/12    
    return np.round(Int_per_card,2)



def active_customers():
    
    active_cust=[]
    total_cust=1000 
    cust_60_days=1/100

    for month in total_months:
        if month>3:
            update_customers=total_cust-cust_60_days*total_cust
            active_cust.append(int(update_customers))    
            total_cust=update_customers
        else:
            active_cust.append(int(total_cust)) 
            
    return active_cust

def int_30days(columns):
    
    int_month=columns[0]
    customers=columns[1]
    avg_day_trans=15 
    billing_period=15 
    
    return np.ceil((avg_day_trans+billing_period+30)/30)*cust_30_days*customers*int_month

def int_45days(columns):
    
    int_month=columns[0]
    customers=columns[1]
    avg_day_trans=15 
    billing_period=15 
    
    return np.ceil((avg_day_trans+billing_period+45)/30)*cust_45_days*customers*int_month

def int_60days(columns):
    
    int_month=columns[0]
    customers=columns[1]
    avg_day_trans=15 
    billing_period=15 
    
    return np.ceil((avg_day_trans+billing_period+60)/30)*cust_60_days*customers*int_month

def total_charge(columns):
    
    return columns[0]+columns[1]+columns[2]+columns[3]    

def PnL_check(): 
    if net_profit_loss>=0:
        return 'Profit margin in percentage'
    else:
        return 'Loss margin in percentage'
    
    

Data_Calc_scenario2['Interest_Relative_to_Balance']=Data_Calc_scenario2['Month_on_Month_Bal'].apply(interest_calc)
Data_Calc_scenario2.insert(4,'Active_customers',active_customers(),True)
Data_Calc_scenario2['Total_Fixed_Late_Fee']=Data_Calc_scenario2['Active_customers'].apply(lambda cust: (1-cust_no_delay)*cust*late_fee)
Data_Calc_scenario2['Total_Interest_30days_late']=Data_Calc_scenario2[['Interest_Relative_to_Balance','Active_customers']].apply(int_30days,axis=1)
Data_Calc_scenario2['Total_Interest_45days_late']=Data_Calc_scenario2[['Interest_Relative_to_Balance','Active_customers']].apply(int_45days,axis=1)
Data_Calc_scenario2['Total_Interest_60days_late']=Data_Calc_scenario2[['Interest_Relative_to_Balance','Active_customers']].apply(int_60days,axis=1)
Data_Calc_scenario2['Total_Late_Charges_Collected']=(Data_Calc_scenario2.iloc[:,5:]).apply(total_charge,axis=1)

Data_Calc_scenario2

Month  Month_on_Month_Bal  Membership_Fee  Interest_Relative_to_Balance  \
       1             1000.00           10.00                         13.22   
       2             1050.00           10.50                         13.88   
       3             1102.50           11.02                         14.58   
       4             1157.62           11.58                         15.30   
       5             1215.51           12.16                         16.07   
       6             1276.28           12.76                         16.87   
       7             1340.10           13.40                         17.72   
       8             1407.10           14.07                         18.60   
       9             1477.46           14.77                         19.53   
      10             1551.33           15.51                         20.51   
      11             1628.89           16.29                         21.54   
      12             1710.34           17.10                         22.61   

   Active_customers  Total_Fixed_Late_Fee  Total_Interest_30days_late  \
               1000               10000.0                   3701.6000   
               1000               10000.0                   3886.4000   
               1000               10000.0                   4082.4000   
                990                9900.0                   4241.1600   
                980                9800.0                   4409.6080   
                970                9700.0                   4581.8920   
                960                9600.0                   4763.1360   
                950                9500.0                   4947.6000   
                941                9410.0                   5145.7644   
                932                9320.0                   5352.2896   
                922                9220.0                   5560.7664   
                913                9130.0                   5780.0204   

   Total_Interest_45days_late  Total_Interest_60days_late  \
                    1983.0000                    396.6000   
                    2082.0000                    416.4000   
                    2187.0000                    437.4000   
                    2272.0500                    454.4100   
                    2362.2900                    472.4580   
                    2454.5850                    490.9170   
                    2551.6800                    510.3360   
                    2650.5000                    530.1000   
                    2756.6595                    551.3319   
                    2867.2980                    573.4596   
                    2978.9820                    595.7964   
                    3096.4395                    619.2879   

   Total_Late_Charges_Collected  
                     16081.2000  
                     16384.8000  
                     16706.8000  
                     16867.6200  
                     17044.3560  
                     17227.3940  
                     17425.1520  
                     17628.2000  
                     17863.7558  
                     18113.0472  
                     18355.5448  
                     18625.7478

# Profit Generated: Scenario 2

In [121]:
loss=0
for month_bal,total_cust in zip(Data_Calc_scenario2['Month_on_Month_Bal'],Data_Calc_scenario2['Active_customers']):
    loss=loss+(loss_rate/12)*total_cust*month_bal  

    
total_mem_fee=0
for mem_fee,total_cust in zip(Data_Calc_scenario2['Membership_Fee'],Data_Calc_scenario2['Active_customers']):
    total_mem_fee=total_mem_fee+mem_fee*total_cust
    
    
total_affil_fee=0
for total_cust in Data_Calc_scenario2['Active_customers']:
    total_affil_fee=total_affil_fee+affil_fee*total_cust
    
    
card_bal=np.sum(Data_Calc_scenario2['Month_on_Month_Bal']*Data_Calc_scenario2['Active_customers'])
    

Expenses=cost_of_funds+operating_cost+loss
Earnings=total_mem_fee+total_affil_fee+np.sum(Data_Calc_scenario2['Total_Late_Charges_Collected'])

margin=Earnings-Expenses 
net_profit_loss=margin/Expenses*100
    
Report=pd.DataFrame({'Total yearly card balance':card_bal,'Earnings':Earnings,'Expenses':Expenses,
              'Margin':margin,PnL_check():net_profit_loss.round(2)},index=[' '])

Report

Total yearly card balance     Earnings      Expenses       Margin  \
                 15250482.62  476397.8876  403126.20655  73271.68105   

   Profit margin in percentage  
                         18.18

Results:
The profit in "Scenario 1" comes out to be ~3.2% and in comparison, 
the profit in "Scenario 2" comes out to be ~18%.

Important Insights and Takeaways:

* The credit card company makes 6 times more profit if the Average balance per account increases 5% monthly in addition to 5% monthly increase in membership fee for 1 year.

* Earnings to expense ratio is: 1.032 in "Scenario 1" and 1.180 in "Scenario 2". Clearly, scenario 2 must be implemented as it can drive higher profits.

* Loss rate (loss incurred due to non payment of bills by customers) increases with increase in average card balance and as a result overall profits are also impacted.

* Interestingly, the more number of customers that defaults or pay after due date, the more profit a company generates in the form of Late fee/Interest charges. So, the company benefits from credit defaults considering 95-99% customers settle the amount.

* Interest when "compunded quarterly and calculated monthly" is higher than the interest when "calculated per day". As per the calculation, It is found that for equal outstanding balance amount and equal number of days taken after due date, there is a difference of ~7% per annum.  